### ISSUE TITLE
Invalidate context after switch user

### ISSUE THREAD

**Issue by onEXHovia** [FriendsOfSymfony/FOSHttpCacheBundle#506]
Currently invalidate executed after logout https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/blob/master/src/Security/Http/Logout/ContextInvalidationSessionLogoutHandler.php. 
I had a problem with feature https://symfony.com/doc/current/security/impersonating_user.html. Should we be invalidation cache after switching?

Example of what I mean

```php
<?php
namespace App\EventSubscriber;

use FOS\HttpCacheBundle\UserContextInvalidator;
use Symfony\Component\EventDispatcher\EventSubscriberInterface;
use Symfony\Component\Security\Http\Event\SwitchUserEvent;
use Symfony\Component\Security\Http\SecurityEvents;

class SwitchUserSubscriber implements EventSubscriberInterface
{
    private $invalidator;

    public function __construct(UserContextInvalidator $invalidator)
    {
        $this->invalidator = $invalidator;
    }

    /**
     * {@inheritdoc}
     */
    public static function getSubscribedEvents(): array
    {
        return [
            SecurityEvents::SWITCH_USER => 'onSwitchUser',
        ];
    }

    public function onSwitchUser(SwitchUserEvent $event): void
    {
        $request = $event->getRequest();
        $this->invalidator->invalidateContext($request->getSession()->getId());
    }
}
```

**Comment by dbu** [comment]
hm, i guess you are right, yes. impersonating a different user will give you a different set of permissions so the cache context hash should not stay the same. thanks for reporting this.

i assume the session stays the same when impersonating, right? and when you "stop impersonating" to become the admin user again, you still keep the ession, but the SwitchUserEvent happens again?

have you tried your proposed solution in your system and seen it working?

could you do a pull request with this? i guess it should be in src/EventListener/SwitchUserListener.php 
can you also add a functional test? maybe a bit along the lines of https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/blob/master/tests/Functional/Security/Http/Logout/ContextInvalidationLogoutHandlerTest.php or however we can make it happen so we can be reasonably certain it works as expected :-)

**Comment by onEXHovia** [comment]
> i assume the session stays the same when impersonating, right? and when you "stop impersonating" to become the admin user again, you still keep the ession, but the SwitchUserEvent happens again?

Quite right,  event `SecurityEvents::SWITCH_USER` happen twice

> have you tried your proposed solution in your system and seen it working?

Yes, used this solution on one of the projects

> could you do a pull request with this? i guess it should be in src/EventListener/SwitchUserListener.php
can you also add a functional test? maybe a bit along the lines of https://github.com/FriendsOfSymfony/FOSHttpCacheBundle/blob/master/tests/Functional/Security/Http/Logout/ContextInvalidationLogoutHandlerTest.php or however we can make it happen so we can be reasonably certain it works as expected :-)

I'm not against :-) in the near future open the PR.

**Comment by dbu** [comment]
fixed in #510
